Option <skip>true|false</skip>

aoterolorenzo opened the pull request:
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk-maven-plugin/blob/master/CONTRIBUTING.md) rules
- [ ] Reviewed by Snyk internal team

Closes #79 

 #### What does this PR do?

Sets both configuration and CLI option to skip the plugin execution.

 #### Where should the reviewer start?

Simple changes in SnykMonitor and SnykTest to add the functionality

 #### How should this be manually tested?

Executing the plugin with the <skip> configuration parameter on the project's pom or
using the CLI option -Dsnyk.skip (just as predisposed in #79's expected behaviour)

 #### Any background context you want to provide?

N/A

 #### What are the relevant tickets?

N/A

 #### Screenshots

N/A

 #### Additional questions

N/A

recena commented on pom.xml:
```diff
@@ -4,7 +4,7 @@
   <groupId>io.snyk</groupId>
   <artifactId>snyk-maven-plugin</artifactId>
-  <version>1.2.8</version>
+  <version>1.2.8-SNAPSHOT</version>
```
I think you should set `1.2.9-SNAPSHOT` version

recena commented on pom.xml:
```diff
@@ -34,8 +34,10 @@
     <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
     <mavenVersion>3.1.0</mavenVersion>
     <aetherVersion>1.1.0</aetherVersion>
+
```
Unrelated change.

recena commented on pom.xml:
```diff
@@ -34,8 +34,10 @@
+
   </properties>
 
+
```
Unrelated change.

recena commented on pom.xml:
```diff
@@ -214,6 +216,7 @@
           <plugin>
             <groupId>org.apache.maven.plugins</groupId>
             <artifactId>maven-gpg-plugin</artifactId>
+            <version>1.6</version>
```
This is an unrelated change but it could be included since improves the readability and avoid problems.

recena commented on src/main/java/io/snyk/maven/plugins/SnykMonitor.java:
```diff
@@ -102,6 +105,12 @@ public void execute() {
     private void executeInternal()
             throws IOException, ParseException, MojoFailureException {
+
+        if(skip) {
```
Please, use the same coding style convention: `if (skip) {`

recena commented on src/main/java/io/snyk/maven/plugins/SnykTest.java:
```diff
@@ -116,10 +123,17 @@ public void execute() throws MojoFailureException {
      */
     private void executeInternal() throws MojoFailureException, IOException {
+
+        if(skip) {
```
Ditto

recena reviewed (changes requested):
I left some comments.

aoterolorenzo commented:
Already passing tests after setting up Api key and endpoint in local except the license-issue. Need of contribution agreement? No incoming sign request not instructions after bring PR from Draft to review.

Output:
`[INFO] --- maven-invoker-plugin:2.0.0:integration-test (integration-test) @ snyk-maven-plugin ---
[INFO] Building: private-repo-module/pom.xml
[INFO] run script verify.bsh
[INFO] ..SUCCESS (3.6 s)
[INFO] Building: license-issues-module/pom.xml
[INFO] run script verify.bsh
**[INFO] ..FAILED (2.2 s)
[INFO]   The post-build script did not succeed. Missing license issue (EPL-1.0) for junit:junit@4.12
[INFO] Building: single-module/pom.xml**
[INFO] run script verify.bsh
[INFO] ..SUCCESS (2.7 s)
[INFO] Building: remediation/pom.xml
[INFO] run script verify.bsh
[INFO] ..SUCCESS (2.3 s)
[INFO] Building: multi-module/pom.xml
[INFO] run script verify.bsh
[INFO] ..SUCCESS (3.5 s)
[INFO] 
[INFO] --- maven-invoker-plugin:2.0.0:verify (integration-test) @ snyk-maven-plugin ---
[INFO] -------------------------------------------------
[INFO] Build Summary:
[INFO]   Passed: 4, Failed: 1, Errors: 0, Skipped: 0
[INFO] -------------------------------------------------
[ERROR] The following builds failed:
[ERROR] *  license-issues-module/pom.xml
[INFO] -------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  17.313 s
[INFO] Finished at: 2020-10-13T13:01:35+02:00
[INFO] ------------------------------------------------------------------------
`

pavel-snyk commented:
Hi @aoterolorenzo, 

can you explain/elaborate your use-case for this option?

If you want to control whether `snyk test/monitor` will be executed during maven build, I would recommend to use Maven build profiles (this is maven-way to enable/disable plugins during builds). Since our plugin execution is not hard-coded in default lifecycle, there is no need to add `skip` option (in compare to `skipTest` option by test plugin).

This is why I would like to know your use-case, perhaps I miss something.

Thanks,
Pavel

pavel-snyk commented:
about failing integration tests: we have a pre-defined configuration for IT-tests. Since envvars will not be exported by GH forks, your configuration is different.
We know it and working on a portable solution.

aoterolorenzo commented:
> Hi @aoterolorenzo,
> 
> can you explain/elaborate your use-case for this option?
> 
> If you want to control whether `snyk test/monitor` will be executed during maven build, I would recommend to use Maven build profiles (this is maven-way to enable/disable plugins during builds). Since our plugin execution is not hard-coded in default lifecycle, there is no need to add `skip` option (in compare to `skipTest` option by test plugin).
> 
> This is why I would like to know your use-case, perhaps I miss something.
> 
> Thanks,
> Pavel

Hi @pavel-snyk,

In general terms, having a skip functionality on the plugin, provides more flexibility when it comes to configure complex multimodule projects.

Certainly you could replicate this behavior with Profiles, but is not a good practice, given the fact that their use in cases related to properties are not recommended by Maven conventions http://maven.apache.org/guides/introduction/introduction-to-profiles.html#profile-pitfalls

Furthermore there’s some quality metric plugins that are currently implemented this feature, like Checkstyle or Sportbugs, which have their respective `skip` functionality. 

Thank you,
Alberto.

recena commented:
I totally agree with @aoterolorenzo. Many Maven Core Plugins provide this mechanism to disable its execution.

aoterolorenzo commented:
@pavel-snyk just checking in, any updates on this?

Thank you,
Alberto

recena commented:
@pavel-snyk Is there any chance to review this proposal?

recena commented:
Hello @pavel-snyk, Is there any way to move forward this proposal?

pavel-snyk commented:
👋 @aoterolorenzo and @recena 
We discussed this proposal internally. That why it took longer:
1) you cannot compare the plugin with maven core plugins, I mentioned it in my comment before. Maven core plugins have builtin lifecycle (defined in xml).
2) about your link and pitfalls for profiles: this is not what the article explains. It means if you have e.g. an appserver with local path, you cannot run a build without defined this path, so the maven build will be not portable.
3) but I see the following use-case: if you have a multi-module project and you defined a snyk execution in parent pom, then disable/enable in child poms. It's doable with profiles or `skip` in configuration. In my opinion profiles are more clean approach, but there are a lot of custom (enterprise specific) builds we want to support.

**Summary**:
- please add a documentation for the `skip` option in readme file: https://github.com/snyk/snyk-maven-plugin#configuration
- take a look for my comment about skip logging

Cheers,
Pavel

recena commented:
@pavel-snyk Thank you so much for your feedback.

> 1. you cannot compare the plugin with maven core plugins, I mentioned it in my comment before. Maven core plugins have builtin lifecycle (defined in xml).

it is not about plugin comparison but use standard behaviors and common configurations.

> 2. about your link and pitfalls for profiles: this is not what the article explains. It means if you have e.g. an appserver with local path, you cannot run a build without defined this path, so the maven build will be not portable.

I agree it is not 100% clear but in my opinion is not good option to use profiles to enable plugins like this. But yes, it is a personal preference.

> 3. but I see the following use-case: if you have a multi-module project and you defined a snyk execution in parent pom, then disable/enable in child poms. It's doable with profiles or `skip` in configuration. In my opinion profiles are more clean approach, but there are a lot of custom (enterprise specific) builds we want to support.

Exactly, this is one of the use cases.
 
> **Summary**:
> 
> * please add a documentation for the `skip` option in readme file: https://github.com/snyk/snyk-maven-plugin#configuration
> * take a look for my comment about skip logging

We don't see that comment.

Regards,

pavel-snyk commented:
> We don't see that comment.
fixed it :), GH requires click on comment button twice 🤦

aoterolorenzo commented:
> > We don't see that comment.
> > fixed it :), GH requires click on comment button twice 🤦

No problem :) Updated in bcea127 addressing your comments. Thanks!

recena reviewed (commented):
IMHO we shoud use more specific log messages.
